src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs

```diff
@@ -483,9 +484,80 @@ private void UpdateCustomColumnsVisibility(ImmutableDictionary<string, Immutable
+                            {
+                                Debug.Fail($"{nameof(SourceReferenceItem.FindUsagesProperties)} has a key '{customColumnName}', but there is no exported '{nameof(AbstractCustomColumnDefinition)}' with this name.");
+                                continue;
+                            }
+
+                            // Ensure that we flip the visibility to false for FSharp and true for other languages.
+                            // Note that the actual UI update happens outside the lock when we
+                            // invoke "TableControl.SetColumnStates" below.
+                            ColumnState2 newColumnStateOpt = null;
+                            if (_customColumnTitleToStatesMap.TryGetValue(customColumnDefinition.Name, out var columnState))
+                            {
+                                newColumnStateOpt = new ColumnState2(columnState.Name, isVisible: false, columnState.Width,
+                                    columnState.SortPriority, columnState.DescendingSort, columnState.GroupingPriority);
+                            }
+                            else
+                            {
+                                newColumnStateOpt = customColumnDefinition.DefaultColumnState;
+                            }
+
+                            if (newColumnStateOpt != null)
+                            {
+                                _customColumnTitleToStatesMap[customColumnDefinition.Name] = newColumnStateOpt;
+
+                                newColumnStates.Add(newColumnStateOpt);
+                            }
+                        }
+                    }
+
+                    // Update the column states if required.
+                    if (newColumnStates.Count > 0)
+                    {
+                        // SetColumnStates API forces a switch to UI thread, so it should be safe to call
+                        // from a background thread here.
+                        // Also note that we will call it only once for each new custom column to add for
+                        // each find references query - the lock above guarantees that newColumnStatesOpt is
+                        // going to be non-null only for the first result that has a non-empty column value.
+                        TableControl.SetColumnStates(newColumnStates.ToImmutable());
+                    }
+                }
+                finally
+                {
+                    newColumnStates.Free();
+                }
+            }
+
             public sealed override Task OnReferenceFoundAsync(SourceReferenceItem reference)
             {
                 UpdateCustomColumnsVisibility(reference.AdditionalPropertiesWithMultipleValues);
+                if (reference.SourceSpan.Document.Project.Language == LanguageNames.FSharp)
+                {
+                    HideContainingMemberAndTypeColumns();
```

Comment:
Doing this for every single reference that is found can be expensive.

Reply:
agreed.  i don't like this at all.

Reply:
i don't even see why we need the special language logic.  can this not just run for all languages and just hide/show these columns if we have any items with those values?